libsel4vm/src/guest_ram.c

```diff
@@ -185,7 +192,11 @@ int vm_ram_find_largest_free_region(vm_t *vm, uintptr_t *addr, size_t *size)
     /* find the largest unallocated region */
     for (int i = 0; i < guest_memory->num_ram_regions; i++) {
         if (!guest_memory->ram_regions[i].allocated &&
-            guest_memory->ram_regions[i].size > largest_size) {
+            guest_memory->ram_regions[i].size > largest_size
+#ifdef CONFIG_X86_64_VTX_64BIT_GUESTS
+            && guest_memory->ram_regions[i].start < UINT32_MAX
```

Comment:
Why does the start address have to be lower than `UINT32_MAX`?

Reply:
The structure used to populate the BIOS uses uint32_t.

```
struct setup_header {
    uint8_t    setup_sects;
    uint16_t   root_flags;
    uint32_t   syssize;
    uint16_t   ram_size;
#define RAMDISK_IMAGE_START_MASK    0x07FF
#define RAMDISK_PROMPT_FLAG     0x8000
#define RAMDISK_LOAD_FLAG       0x4000
    uint16_t   vid_mode;
    uint16_t   root_dev;
    uint16_t   boot_flag;
    uint16_t   jump;
    uint32_t   header;
    uint16_t   version;
    uint32_t   realmode_swtch;
    uint16_t   start_sys;
    uint16_t   kernel_version;
    uint8_t    type_of_loader;
    uint8_t    loadflags;
#define LOADED_HIGH (1<<0)
#define QUIET_FLAG  (1<<5)
#define KEEP_SEGMENTS   (1<<6)
#define CAN_USE_HEAP    (1<<7)
    uint16_t   setup_move_size;
    uint32_t   code32_start;
    uint32_t   ramdisk_image;
    uint32_t   ramdisk_size;
```

Reply:
But why check the start, surely the end must be `< UINT32_MAX`?

Reply:
The function just checks for the largest free, contiguous region. My guess is that if a region started in 32-bit space, and ended in 64-bit space, that would be fine, since only the start address matters for Linux, and there's only a few hundred bytes of code before the processor state is switched to 64-bit mode.

However, we can check for `.start + .size` if you'd prefer.

Reply:
Well, in that case you should check for `<= UINT32_MAX` instead.

I'm fine with whatever works, but `+ .size` feels more correct to me.